### PR TITLE
Implement engine_getBlobsV1

### PR DESCRIPTION
**Pull request by michaelsproul**
Implement the new API for fetching blobs from the mempool to improve propagation:

- https://github.com/ethereum/execution-apis/pull/559

### CODE REVIEW

**Comment by Rjected** [comment]
just noting that requesting by commitment or versioned hash will require refactoring our on disk blob store to work with those types

**Comment by michaelsproul** [comment]
Thanks for the feedback @Rjected

Maybe the on-disk blob store could keep a mapping from versioned hash to transaction hash(s)? It could build the mapping at startup and keep it in memory? Or it could write the mapping to disk and read/write from it on-demand. My intuition says the in-memory mapping would be easier because there aren't too many blobs, but you'd be a better judge of whether that's feasible.

I'm guessing the `DiskFileBlobStore` here is what gets used in prod?

https://github.com/michaelsproul/reth/blob/de6332b51e9d4bdf9d3e75a1523ff1cdfb862686/crates/transaction-pool/src/blobstore/disk.rs#L252-L271

**Comment by Rjected** [comment]
> Thanks for the feedback @Rjected
> 
> Maybe the on-disk blob store could keep a mapping from versioned hash to transaction hash(s)? It could build the mapping at startup and keep it in memory? Or it could write the mapping to disk and read/write from it on-demand. My intuition says the in-memory mapping would be easier because there aren't too many blobs, but you'd be a better judge of whether that's feasible.
> 
> I'm guessing the `DiskFileBlobStore` here is what gets used in prod?
> 
> https://github.com/michaelsproul/reth/blob/de6332b51e9d4bdf9d3e75a1523ff1cdfb862686/crates/transaction-pool/src/blobstore/disk.rs#L252-L271

Yeah, the `DiskFileBlobStore` is used in practice. Here is where the blob is primarily stored in the blob store:
https://github.com/paradigmxyz/reth/blob/6e146e1140e874b88c304fdc3c5ac0e046e66263/crates/transaction-pool/src/pool/mod.rs#L1188-L1191
https://github.com/paradigmxyz/reth/blob/6e146e1140e874b88c304fdc3c5ac0e046e66263/crates/transaction-pool/src/pool/mod.rs#L788-L795
 
And here is where it is retrieved:
https://github.com/paradigmxyz/reth/blob/6e146e1140e874b88c304fdc3c5ac0e046e66263/crates/transaction-pool/src/pool/mod.rs#L299-L308

Deletion gets slightly more tricky, since we use `BlobStoreUpdates::Finalized` to let the blob store know it should remove a set of blob files.

Replacement also complicates things, we would need to surface the versioned hash there:
https://github.com/paradigmxyz/reth/blob/6e146e1140e874b88c304fdc3c5ac0e046e66263/crates/transaction-pool/src/pool/mod.rs#L460-L463

So yeah, seems like a lot of things to refactor if we wanted to change the key to versioned hash entirely, although it might make sense. The in memory mapping does seem like the quickest way to implement this

**Comment by mattsse** [comment]
another approach that could work for the request, is
fetch all 4844 transactions from the pool, this shouldn't be too expensive because `Vec<Arc` and then simply try finding the matching tx and finally fetch the blob by txhash

**Comment by michaelsproul** [comment]
@mattsse I've updated the prototype to follow the latest spec by doing something similar with the in-memory blob cache (just iterate the full cache). I think the main inefficiency is probably recalculating the versioned hash, but that could be optimised out

File: crates/primitives/src/transaction/mod.rs [hunk 1]
```diff
@@ -415,7 +415,7 @@ impl Transaction {
```
Comment (michaelsproul) [inline crates/primitives/src/transaction/mod.rs]: I need to revert the fmt changes to this file

**Comment by Rjected** [comment]
@michaelsproul do you mind opening this against the main reth repo? It will be easier for the team to track

**Comment by michaelsproul** [comment]
Good idea. Closing in favour of:

- https://github.com/paradigmxyz/reth/pull/9723

**Pull request closed without merging**